### ISSUE TITLE
Allow things to be renamed in the bindings config

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#1426]
We should allow functions, types, and other objects to be renamed in the bindings config, overriding the generated names.

The biggest use case for this is Swift, which doesn't support namespaces.  This means that if you UniFFI a collection of rust crates, the names of each type needs to be unique or else you can name collisions -- for example if 2 crates defined an `Error` type and you tried to import both generated modules.

This can be worked around by keeping each type name unique across all the UDL files, but that's not so idiomatic for languages whose modules form namespaces -- For example in Python `foo.Error` and `bar.Error` is usually more natural than `foo.FooError` and `bar.BarError`.

This means that the in many cases it would be good to allow types to share a name, but rename them for Swift to add a prefix.  I think the renames could be specified in `uniffi.toml` like this:

```
[bindings.swift.names]
SomeType = "RenamedType"
some_function = "renamed_function"
```

**Comment by arg0d** [comment]
Our Apple team is also interesting in something like this. They would like to use Swift Uniffi interface throughout their code base. However, they have their own interface naming convention, so they would like generated Uniffi interfaces to be named according to their own interface naming convention. Right now the only way to accomplish this would be to fork Uniffi and change the interface naming in template files directly.

**Comment by kriswuollett** [comment]
> I think the renames could be specified in `uniffi.toml` like this:
> 
> ```
> [bindings.swift.names]
> SomeType = "RenamedType"
> some_function = "renamed_function"
> ```

While I wouldn't want to discourage support the option of renaming individual types, the resulting config files may likely become very unwieldy if renaming things one at a time. I'd assume the most common case for Swift would be to namespace with a **per-crate prefix for everything** like back in the days of Objective-C. The issues with the above are:

1.  Need to fully qualify types potentially due to name conflicts with transitive dependencies, **and that** quoted identifiers/keys are needed since [punctuation isn't allowed in raw keys](https://github.com/toml-lang/toml/blob/dc5af251cd348ee053d9355359743767e3d0b123/toml.md?plain=1#L108), but this seems unavoidable in TOML,  e.g.:

```toml
[bindings.swift.names]
"my_crate::foo::SomeType" = "renamedType"
```

2. Transitive dependencies if renaming is supported also may cause confusion of what type name to use as the original when renaming. Do you always have to check the overrides in config files throughout the dependency chain? Reminds me of Kustomize on how [they allow matching any of the previously used names for an entity](https://github.com/kubernetes-sigs/kustomize/blob/214aa2aae2ab07f8c95be1a32f7ab6461fb08ebe/examples/transformerconfigs/README.md?plain=1#L180-L189). Sounds error-prone.

So what may be very useful and easy to use, at least as far as a language like Swift is concerned is a two-level config:

```toml
[bindings.swift.package.prefixes]
"my_crate" = "mc"
"my_2nd_crate" = "m2c"
"my_3rd_crate::baz" = "m3c_bz"
```

This would have the following effects:

```
my_crate::hello_world (fn) -> mcHelloWorld
my_crate::Greeeter (struct) -> MCGreeter
my_2nd_crate::Greeter (struct) -> M2CGreeter
my_3rd_crate::Greeter (struct) -> My3rdCrateGreeter (note my_3rd_crate wasn't mapped, so default applied)
my_3rd_crate::baz::Greeter (struct) -> M3CBzGreeter
my_3rd_crate::baz::bat::Greeter (struct) -> M3CBzBatGreeter
my_3rd_crate::baz::bat::hello_world (fn) -> m3cBzBatHelloWorld
```

Basically assume everything is named according to Rust conventions. Use the Rust conventions to rename based on the fully qualified name into the target language conventions. Allow per-crate mappings that are inherited by child crates to help make things unwieldy like `My3rdCrateBazBatGreeter` instead of the example above `M3CBzBatGreeter`. Finally add support for individual renames in case the above policy doesn't work for an individual type/developer.

**Comment by kriswuollett** [comment]
Probably also need a config option to create type aliases for the original name with a deprecation/todo warning so that large code bases can migrate to a new naming policy gradually.

**Comment by data-sync-user** [comment]
➤ Ben Dean-Kawamura commented:

This issue is for Python only.  We can tackle Swift and Kotlin later.
